File: android/src/main/java/expo/modules/libsignal/SessionOps.kt
package expo.modules.libsignal

import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.time.Instant
import org.signal.libsignal.protocol.IdentityKey
import org.signal.libsignal.protocol.IdentityKeyPair as SignalIdentityKeyPair
import org.signal.libsignal.protocol.SessionBuilder
import org.signal.libsignal.protocol.SessionCipher
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.message.CiphertextMessage
import org.signal.libsignal.protocol.message.PreKeySignalMessage
import org.signal.libsignal.protocol.message.SignalMessage
import org.signal.libsignal.protocol.state.KyberPreKeyRecord
import org.signal.libsignal.protocol.state.PreKeyRecord
import org.signal.libsignal.protocol.state.SessionRecord
import org.signal.libsignal.protocol.state.SignedPreKeyRecord
import org.signal.libsignal.protocol.state.impl.InMemorySignalProtocolStore

// The op boundary: argument Records carry only primitives (Expo Modules on
// Android converts incoming Records via plain maps, which cannot carry
// SharedObjects or typed arrays), every byte payload is a positional
// Uint8Array argument (max 8 args per function — decryptPreKeySignalOp uses
// all 8), and PreKeyBundle — which has no serialized form in libsignal — is
// a positional SharedObject. Result Records may carry bytes: the return path
// converts ByteArray fields correctly on both platforms.

class SessionOpConfig : Record {
  @Field var remoteName: String = ""
  @Field var remoteDeviceId: Int = 0
  @Field var localName: String = ""
  @Field var localDeviceId: Int = 0
  @Field var ourRegistrationId: Int = 0
  @Field var nowMs: Double = 0.0
}

class ProcessPreKeyBundleResult : Record {
  @Field var newSession: ByteArray = ByteArray(0)
  @Field var identityChange: String = "newOrUnchanged"
  @Field var trustedRemoteIdentity: ByteArray = ByteArray(0)
}

class EncryptResult : Record {
  @Field var messageType: String = ""
  @Field var preKeySignalMessage: ByteArray? = null
  @Field var signalMessage: ByteArray? = null
  @Field var newSession: ByteArray = ByteArray(0)
  @Field var identityChange: String? = null
}

class DecryptPreKeySignalResult : Record {
  @Field var plaintext: ByteArray = ByteArray(0)
  @Field var newSession: ByteArray = ByteArray(0)
  @Field var identityChange: String? = null
  @Field var consumedPreKeyId: Int? = null
  @Field var kyberPreKeyId: Int = 0
}

class DecryptSignalResult : Record {
  @Field var plaintext: ByteArray = ByteArray(0)
  @Field var newSession: ByteArray = ByteArray(0)
  @Field var identityChange: String? = null
}

internal class ParsedSessionOpArgs(
  val remoteAddress: SignalProtocolAddress,
  val localAddress: SignalProtocolAddress,
  val identityKeyPair: SignalIdentityKeyPair,
  val existingSession: SessionRecord?,
  val existingRemoteIdentity: IdentityKey?,
)

internal fun parseSessionOpArgs(
  config: SessionOpConfig,
  ourIdentityKeyPair: ByteArray,
  existingSession: ByteArray?,
  existingRemoteIdentity: ByteArray?,
): ParsedSessionOpArgs {
  return ParsedSessionOpArgs(
    remoteAddress = SignalProtocolAddress(config.remoteName, config.remoteDeviceId),
    localAddress = SignalProtocolAddress(config.localName, config.localDeviceId),
    identityKeyPair = SignalIdentityKeyPair(ourIdentityKeyPair),
    existingSession = existingSession?.let { SessionRecord(it) },
    existingRemoteIdentity = existingRemoteIdentity?.let { IdentityKey(it) },
  )
}

internal fun seedStore(
  identity: SignalIdentityKeyPair,
  registrationId: Int,
  remoteAddress: SignalProtocolAddress? = null,
  existingSession: SessionRecord? = null,
  existingRemoteIdentity: IdentityKey? = null,
): InMemorySignalProtocolStore {
  val store = InMemorySignalProtocolStore(identity, registrationId)
  if (existingSession != null && remoteAddress != null) {
    store.storeSession(remoteAddress, existingSession)
  }
  if (existingRemoteIdentity != null && remoteAddress != null) {
    store.saveIdentity(remoteAddress, existingRemoteIdentity)
  }
  return store
}

internal fun identityChangeString(
  store: InMemorySignalProtocolStore,
  remoteAddress: SignalProtocolAddress,
  existing: IdentityKey?,
): String {
  val now = store.getIdentity(remoteAddress)
  if (now != null && existing != null && now == existing) {
    return "newOrUnchanged"
  }
  return if (existing == null) "newOrUnchanged" else "replacedExisting"
}

internal fun runProcessPreKeyBundleOp(
  config: SessionOpConfig,
  bundle: PreKeyBundleRef,
  ourIdentityKeyPair: ByteArray,
  existingSession: ByteArray?,
  existingRemoteIdentity: ByteArray?,
): ProcessPreKeyBundleResult {
  val parsed = parseSessionOpArgs(config, ourIdentityKeyPair, existingSession, existingRemoteIdentity)

  val store = seedStore(
    identity = parsed.identityKeyPair,
    registrationId = config.ourRegistrationId,
    remoteAddress = parsed.remoteAddress,
    existingSession = parsed.existingSession,
    existingRemoteIdentity = parsed.existingRemoteIdentity,
  )

  val builder = SessionBuilder(store, store, store, store, parsed.remoteAddress, parsed.localAddress)
  builder.process(bundle.bundle, Instant.ofEpochMilli(config.nowMs.toLong()))

  val newSession = store.loadSession(parsed.remoteAddress)
    ?: throw IllegalStateException("processPreKeyBundle did not produce a session")
  val trustedRemote = store.getIdentity(parsed.remoteAddress) ?: bundle.bundle.identityKey

  val result = ProcessPreKeyBundleResult()
  result.newSession = newSession.serialize()
  result.identityChange = identityChangeString(store, parsed.remoteAddress, parsed.existingRemoteIdentity)
  result.trustedRemoteIdentity = trustedRemote.serialize()
  return result
}

internal fun runEncryptOp(
  config: SessionOpConfig,
  plaintext: ByteArray,
  ourIdentityKeyPair: ByteArray,
  existingSession: ByteArray,
  remoteIdentity: ByteArray?,
): EncryptResult {
  val parsed = parseSessionOpArgs(config, ourIdentityKeyPair, existingSession, remoteIdentity)

  val store = seedStore(
    identity = parsed.identityKeyPair,
    registrationId = config.ourRegistrationId,
    remoteAddress = parsed.remoteAddress,
    existingSession = parsed.existingSession,
    existingRemoteIdentity = parsed.existingRemoteIdentity,
  )

  // Note the upstream asymmetry: SessionBuilder's constructor takes
  // (remoteAddress, localAddress) but SessionCipher's takes
  // (localAddress, remoteAddress).
  val cipher = SessionCipher(store, store, store, store, store, parsed.localAddress, parsed.remoteAddress)
  val ciphertext = cipher.encrypt(plaintext, Instant.ofEpochMilli(config.nowMs.toLong()))

  val newSession = store.loadSession(parsed.remoteAddress)
    ?: throw IllegalStateException("encryptOp produced no session")

  val result = EncryptResult()
  result.newSession = newSession.serialize()
  result.identityChange = if (parsed.existingRemoteIdentity == null) null else "newOrUnchanged"

  when (ciphertext.type) {
    CiphertextMessage.PREKEY_TYPE -> {
      result.messageType = "preKeySignal"
      result.preKeySignalMessage = ciphertext.serialize()
    }
    CiphertextMessage.WHISPER_TYPE -> {
      result.messageType = "signal"
      result.signalMessage = ciphertext.serialize()
    }
    else -> throw IllegalStateException("encryptOp produced unexpected ciphertext type ${ciphertext.type}")
  }
  return result
}

internal fun runDecryptPreKeySignalOp(
  config: SessionOpConfig,
  message: ByteArray,
  ourIdentityKeyPair: ByteArray,
  existingSession: ByteArray?,
  existingRemoteIdentity: ByteArray?,
  preKey: ByteArray?,
  signedPreKey: ByteArray,
  kyberPreKey: ByteArray,
): DecryptPreKeySignalResult {
  val parsed = parseSessionOpArgs(config, ourIdentityKeyPair, existingSession, existingRemoteIdentity)

  val msg = PreKeySignalMessage(message)
  val parsedSignedPreKey = SignedPreKeyRecord(signedPreKey)
  val parsedKyberPreKey = KyberPreKeyRecord(kyberPreKey)

  val store = seedStore(
    identity = parsed.identityKeyPair,
    registrationId = config.ourRegistrationId,
    remoteAddress = parsed.remoteAddress,
    existingSession = parsed.existingSession,
    existingRemoteIdentity = parsed.existingRemoteIdentity,
  )

  preKey?.let {
    val parsedPreKey = PreKeyRecord(it)
    store.storePreKey(parsedPreKey.id, parsedPreKey)
  }
  store.storeSignedPreKey(parsedSignedPreKey.id, parsedSignedPreKey)
  store.storeKyberPreKey(parsedKyberPreKey.id, parsedKyberPreKey)

  // Note the upstream asymmetry: SessionBuilder's constructor takes
  // (remoteAddress, localAddress) but SessionCipher's takes
  // (localAddress, remoteAddress).
  val cipher = SessionCipher(store, store, store, store, store, parsed.localAddress, parsed.remoteAddress)
  val plaintext = cipher.decrypt(msg)

  val newSession = store.loadSession(parsed.remoteAddress)
    ?: throw IllegalStateException("decryptPreKeySignalOp produced no session")

  val msgPreKeyId = msg.preKeyId
  val consumed = if (msgPreKeyId.isPresent) msgPreKeyId.get() else null

  val result = DecryptPreKeySignalResult()
  result.plaintext = plaintext
  result.newSession = newSession.serialize()
  result.identityChange = identityChangeString(store, parsed.remoteAddress, parsed.existingRemoteIdentity)
  result.consumedPreKeyId = consumed
  result.kyberPreKeyId = parsedKyberPreKey.id
  return result
}

internal fun runDecryptSignalOp(
  config: SessionOpConfig,
  message: ByteArray,
  ourIdentityKeyPair: ByteArray,
  existingSession: ByteArray,
  remoteIdentity: ByteArray?,
): DecryptSignalResult {
  val parsed = parseSessionOpArgs(config, ourIdentityKeyPair, existingSession, remoteIdentity)

  val msg = SignalMessage(message)

  val store = seedStore(
    identity = parsed.identityKeyPair,
    registrationId = config.ourRegistrationId,
    remoteAddress = parsed.remoteAddress,
    existingSession = parsed.existingSession,
    existingRemoteIdentity = parsed.existingRemoteIdentity,
  )

  // Note the upstream asymmetry: SessionBuilder's constructor takes
  // (remoteAddress, localAddress) but SessionCipher's takes
  // (localAddress, remoteAddress).
  val cipher = SessionCipher(store, store, store, store, store, parsed.localAddress, parsed.remoteAddress)
  val plaintext = cipher.decrypt(msg)

  val newSession = store.loadSession(parsed.remoteAddress)
    ?: throw IllegalStateException("decryptSignalOp produced no session")

  val result = DecryptSignalResult()
  result.plaintext = plaintext
  result.newSession = newSession.serialize()
  result.identityChange = identityChangeString(store, parsed.remoteAddress, parsed.existingRemoteIdentity)
  return result
}

File: ios/SessionOps.swift
import Foundation
import ExpoModulesCore
import LibSignalClient

// The op boundary: argument Records carry only primitives (Expo Modules on
// Android converts incoming Records via plain maps, which cannot carry
// SharedObjects or typed arrays), every byte payload is a positional
// Uint8Array argument (max 8 args per function — decryptPreKeySignalOp uses
// all 8), and PreKeyBundle — which has no serialized form in libsignal — is
// a positional SharedObject. Result Records may carry bytes: the return path
// converts ByteArray/Data fields correctly on both platforms.

// MARK: - Records

struct SessionOpConfig: Record {
  @Field var remoteName: String = ""
  @Field var remoteDeviceId: UInt32 = 0
  @Field var localName: String = ""
  @Field var localDeviceId: UInt32 = 0
  @Field var ourRegistrationId: UInt32 = 0
  @Field var nowMs: Double = 0
}

struct ProcessPreKeyBundleResult: Record {
  @Field var newSession: Data = Data()
  @Field var identityChange: String = ""
  @Field var trustedRemoteIdentity: Data = Data()
}

struct EncryptResult: Record {
  @Field var messageType: String = ""
  @Field var preKeySignalMessage: Data? = nil
  @Field var signalMessage: Data? = nil
  @Field var newSession: Data = Data()
  @Field var identityChange: String? = nil
}

struct DecryptPreKeySignalResult: Record {
  @Field var plaintext: Data = Data()
  @Field var newSession: Data = Data()
  @Field var identityChange: String? = nil
  @Field var consumedPreKeyId: UInt32? = nil
  @Field var kyberPreKeyId: UInt32? = nil
}

struct DecryptSignalResult: Record {
  @Field var plaintext: Data = Data()
  @Field var newSession: Data = Data()
  @Field var identityChange: String? = nil
}

// MARK: - Helpers

private struct ParsedSessionOpArgs {
  let remoteAddress: ProtocolAddress
  let localAddress: ProtocolAddress
  let identityKeyPair: IdentityKeyPair
  let existingSession: SessionRecord?
  let existingRemoteIdentity: IdentityKey?
}

private func parseSessionOpArgs(
  config: SessionOpConfig,
  ourIdentityKeyPair: Data,
  existingSession: Data?,
  existingRemoteIdentity: Data?
) throws -> ParsedSessionOpArgs {
  return ParsedSessionOpArgs(
    remoteAddress: try ProtocolAddress(name: config.remoteName, deviceId: config.remoteDeviceId),
    localAddress: try ProtocolAddress(name: config.localName, deviceId: config.localDeviceId),
    identityKeyPair: try IdentityKeyPair(bytes: ourIdentityKeyPair),
    existingSession: try existingSession.map { try SessionRecord(bytes: $0) },
    existingRemoteIdentity: try existingRemoteIdentity.map { try IdentityKey(bytes: $0) }
  )
}

func seedStore(
  identityKeyPair: IdentityKeyPair,
  registrationId: UInt32,
  remoteAddress: ProtocolAddress? = nil,
  existingSession: SessionRecord? = nil,
  existingRemoteIdentity: IdentityKey? = nil
) throws -> InMemorySignalProtocolStore {
  let store = InMemorySignalProtocolStore(identity: identityKeyPair, registrationId: registrationId)
  let ctx = NullContext()
  if let session = existingSession, let addr = remoteAddress {
    try store.storeSession(session, for: addr, context: ctx)
  }
  if let ident = existingRemoteIdentity, let addr = remoteAddress {
    _ = try store.saveIdentity(ident, for: addr, context: ctx)
  }
  return store
}

// Captures which kyber prekey libsignal marks used during PQ decrypt — the
// message does not expose the id in 0.94.4, so this callback is the only
// place the real id surfaces.
final class RecordingSignalProtocolStore: InMemorySignalProtocolStore {
  var usedKyberPreKeyId: UInt32?

  override func markKyberPreKeyUsed(id: UInt32, signedPreKeyId: UInt32, baseKey: PublicKey, context: StoreContext) throws {
    usedKyberPreKeyId = id
    try super.markKyberPreKeyUsed(id: id, signedPreKeyId: signedPreKeyId, baseKey: baseKey, context: context)
  }
}

// Mirror of src/core/recordList.ts: big-endian u32 length prefix per record.
func decodeRecordList(_ blob: Data) throws -> [Data] {
  let bytes = [UInt8](blob)
  var records: [Data] = []
  var offset = 0
  while offset < bytes.count {
    guard offset + 4 <= bytes.count else {
      throw Exception(name: "LibsignalError", description: "recordList: truncated length prefix")
    }
    let len = (Int(bytes[offset]) << 24) | (Int(bytes[offset + 1]) << 16) | (Int(bytes[offset + 2]) << 8) | Int(bytes[offset + 3])
    offset += 4
    guard offset + len <= bytes.count else {
      throw Exception(name: "LibsignalError", description: "recordList: truncated record")
    }
    records.append(Data(bytes[offset..<(offset + len)]))
    offset += len
  }
  return records
}

func identityChangeString(
  store: InMemorySignalProtocolStore,
  remoteAddress: ProtocolAddress,
  existing: IdentityKey?
) throws -> String {
  let now = try store.identity(for: remoteAddress, context: NullContext())
  if let now = now, let existing = existing, now == existing {
    return "newOrUnchanged"
  }
  return existing == nil ? "newOrUnchanged" : "replacedExisting"
}

// MARK: - processPreKeyBundleOp

func runProcessPreKeyBundleOp(
  config: SessionOpConfig,
  bundle: PreKeyBundleRef,
  ourIdentityKeyPair: Data,
  existingSession: Data?,
  existingRemoteIdentity: Data?
) throws -> ProcessPreKeyBundleResult {
  let parsed = try parseSessionOpArgs(
    config: config,
    ourIdentityKeyPair: ourIdentityKeyPair,
    existingSession: existingSession,
    existingRemoteIdentity: existingRemoteIdentity
  )

  let ctx = NullContext()
  let store = try seedStore(
    identityKeyPair: parsed.identityKeyPair,
    registrationId: config.ourRegistrationId,
    remoteAddress: parsed.remoteAddress,
    existingSession: parsed.existingSession,
    existingRemoteIdentity: parsed.existingRemoteIdentity
  )

  try processPreKeyBundle(
    bundle.bundle,
    for: parsed.remoteAddress,
    ourAddress: parsed.localAddress,
    sessionStore: store,
    identityStore: store,
    now: Date(timeIntervalSince1970: config.nowMs / 1000.0),
    context: ctx
  )

  guard let newSession = try store.loadSession(for: parsed.remoteAddress, context: ctx) else {
    throw Exception(name: "LibsignalError", description: "processPreKeyBundle did not produce a session")
  }
  let trustedRemoteIdentity = try store.identity(for: parsed.remoteAddress, context: ctx)
    ?? bundle.bundle.identityKey
  let change = try identityChangeString(
    store: store,
    remoteAddress: parsed.remoteAddress,
    existing: parsed.existingRemoteIdentity
  )

  var result = ProcessPreKeyBundleResult()
  result.newSession = newSession.serialize()
  result.identityChange = change
  result.trustedRemoteIdentity = trustedRemoteIdentity.serialize()
  return result
}

// MARK: - encryptOp

func runEncryptOp(
  config: SessionOpConfig,
  plaintext: Data,
  ourIdentityKeyPair: Data,
  existingSession: Data,
  remoteIdentity: Data?
) throws -> EncryptResult {
  let parsed = try parseSessionOpArgs(
    config: config,
    ourIdentityKeyPair: ourIdentityKeyPair,
    existingSession: existingSession,
    existingRemoteIdentity: remoteIdentity
  )

  let ctx = NullContext()
  let store = try seedStore(
    identityKeyPair: parsed.identityKeyPair,
    registrationId: config.ourRegistrationId,
    remoteAddress: parsed.remoteAddress,
    existingSession: parsed.existingSession,
    existingRemoteIdentity: parsed.existingRemoteIdentity
  )

  let ciphertext = try signalEncrypt(
    message: plaintext,
    for: parsed.remoteAddress,
    localAddress: parsed.localAddress,
    sessionStore: store,
    identityStore: store,
    now: Date(timeIntervalSince1970: config.nowMs / 1000.0),
    context: ctx
  )

  guard let newSession = try store.loadSession(for: parsed.remoteAddress, context: ctx) else {
    throw Exception(name: "LibsignalError", description: "encryptOp produced no session")
  }

  var result = EncryptResult()
  result.newSession = newSession.serialize()
  result.identityChange = "newOrUnchanged"

  switch ciphertext.messageType {
  case .preKey:
    result.messageType = "preKeySignal"
    result.preKeySignalMessage = Data(ciphertext.serialize())
  case .whisper:
    result.messageType = "signal"
    result.signalMessage = Data(ciphertext.serialize())
  default:
    throw Exception(name: "LibsignalError", description: "encryptOp produced unexpected ciphertext type \(ciphertext.messageType.rawValue)")
  }
  return result
}

// MARK: - decryptPreKeySignalOp

func runDecryptPreKeySignalOp(
  config: SessionOpConfig,
  message: Data,
  ourIdentityKeyPair: Data,
  existingSession: Data?,
  existingRemoteIdentity: Data?,
  preKey: Data?,
  signedPreKey: Data,
  kyberPreKeys: Data
) throws -> DecryptPreKeySignalResult {
  let parsed = try parseSessionOpArgs(
    config: config,
    ourIdentityKeyPair: ourIdentityKeyPair,
    existingSession: existingSession,
    existingRemoteIdentity: existingRemoteIdentity
  )

  let parsedMessage = try PreKeySignalMessage(bytes: message)
  let parsedSignedPreKey = try SignedPreKeyRecord(bytes: signedPreKey)

  let ctx = NullContext()
  let store = RecordingSignalProtocolStore(
    identity: parsed.identityKeyPair,
    registrationId: config.ourRegistrationId
  )
  if let session = parsed.existingSession {
    try store.storeSession(session, for: parsed.remoteAddress, context: ctx)
  }
  if let ident = parsed.existingRemoteIdentity {
    _ = try store.saveIdentity(ident, for: parsed.remoteAddress, context: ctx)
  }

  if let preKeyData = preKey {
    let parsedPreKey = try PreKeyRecord(bytes: preKeyData)
    try store.storePreKey(parsedPreKey, id: parsedPreKey.id, context: ctx)
  }
  try store.storeSignedPreKey(parsedSignedPreKey, id: parsedSignedPreKey.id, context: ctx)
  for recordBytes in try decodeRecordList(kyberPreKeys) {
    let record = try KyberPreKeyRecord(bytes: recordBytes)
    try store.storeKyberPreKey(record, id: record.id, context: ctx)
  }

  // Read consumedPreKeyId before decrypt (the message carries it; libsignal removes the record during decrypt).
  let consumedPreKeyId: UInt32? = try parsedMessage.preKeyId()

  let plaintext = try signalDecryptPreKey(
    message: parsedMessage,
    from: parsed.remoteAddress,
    localAddress: parsed.localAddress,
    sessionStore: store,
    identityStore: store,
    preKeyStore: store,
    signedPreKeyStore: store,
    kyberPreKeyStore: store,
    context: ctx
  )

  guard let newSession = try store.loadSession(for: parsed.remoteAddress, context: ctx) else {
    throw Exception(name: "LibsignalError", description: "decryptPreKeySignalOp produced no session")
  }

  var result = DecryptPreKeySignalResult()
  result.plaintext = Data(plaintext)
  result.newSession = newSession.serialize()
  result.identityChange = try identityChangeString(
    store: store,
    remoteAddress: parsed.remoteAddress,
    existing: parsed.existingRemoteIdentity
  )
  result.consumedPreKeyId = consumedPreKeyId
  result.kyberPreKeyId = store.usedKyberPreKeyId
  return result
}

// MARK: - decryptSignalOp

func runDecryptSignalOp(
  config: SessionOpConfig,
  message: Data,
  ourIdentityKeyPair: Data,
  existingSession: Data,
  remoteIdentity: Data?
) throws -> DecryptSignalResult {
  let parsed = try parseSessionOpArgs(
    config: config,
    ourIdentityKeyPair: ourIdentityKeyPair,
    existingSession: existingSession,
    existingRemoteIdentity: remoteIdentity
  )

  let parsedMessage = try SignalMessage(bytes: message)

  let ctx = NullContext()
  let store = try seedStore(
    identityKeyPair: parsed.identityKeyPair,
    registrationId: config.ourRegistrationId,
    remoteAddress: parsed.remoteAddress,
    existingSession: parsed.existingSession,
    existingRemoteIdentity: parsed.existingRemoteIdentity
  )

  let plaintext = try signalDecrypt(
    message: parsedMessage,
    from: parsed.remoteAddress,
    to: parsed.localAddress,
    sessionStore: store,
    identityStore: store,
    context: ctx
  )

  guard let newSession = try store.loadSession(for: parsed.remoteAddress, context: ctx) else {
    throw Exception(name: "LibsignalError", description: "decryptSignalOp produced no session")
  }

  var result = DecryptSignalResult()
  result.plaintext = Data(plaintext)
  result.newSession = newSession.serialize()
  result.identityChange = try identityChangeString(
    store: store,
    remoteAddress: parsed.remoteAddress,
    existing: parsed.existingRemoteIdentity
  )
  return result
}
